packages/shared/src/beacon.ts

```diff
@@ -0,0 +1,105 @@
+/* -----------------------------------
+ *
+ * sendBeacon with XMLHttpRequest
+ *
+ * -------------------------------- */
+
+const syncEvents = new Set(['unload', 'beforeunload', 'pagehide']);
+
+function sendBeaconXHR(url: string | URL, data?: XMLHttpRequestBodyInit | null): boolean {
+   const eventType = this?.event?.type;
+   const sync = syncEvents.has((eventType || '').toLowerCase());
+
+   const xhr = new XMLHttpRequest();
+   xhr.open('POST', url, !sync);
+   xhr.withCredentials = true;
+   xhr.setRequestHeader('Accept', '*/*');
+
+
+   if (typeof data === 'string') {
+      xhr.setRequestHeader('Content-Type', 'text/plain;charset=UTF-8');
+      xhr.responseType = 'text';
+   } else if ((data instanceof Blob) && data.type) {
+      xhr.setRequestHeader('Content-Type', data.type);
+   }
+
+   try {
+      xhr.send(data);
+   } catch (error) {
+      return false;
+   }
+
+   return true;
+}
+
+/* -----------------------------------
+ *
+ * sendBeacon with Fetch
+ *
+ * -------------------------------- */
+
+function sendBeaconFetch(url: string | URL, data?: XMLHttpRequestBodyInit | null): boolean {
+   try {
+      fetch(url, {
+         method: 'POST',
+         body: data,
+         credentials: 'include',
+         mode: 'cors',
+         keepalive: true, // Not present on Firefox
+      })
+      .catch(() => { }); // Avoid unhandledrejection
+      return true;
+   } catch (e) {
+      return false;
+   }
+}
+
+/* -----------------------------------
+ *
+ * sendBeacon
+ *
+ * -------------------------------- */
+
+function sendBeacon(url: string | URL, data?: XMLHttpRequestBodyInit | null): boolean {
+   // #1: navigator.sendBeacon (window Scope)
+   const hasBeaconApi = (typeof navigator !== 'undefined' && typeof navigator.sendBeacon === 'function');
+   if (hasBeaconApi) {
+      if (data) {
+         return navigator.sendBeacon(url, data);
+      }
+      return navigator.sendBeacon(url);
+   }
+
+   // #2: XMLHttpRequest (Worker Scope, except ServiceWorker)
+   // https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#sect1
+   const hasXHR = (typeof XMLHttpRequest !== 'undefined');
+   if (hasXHR) {
+      if (data) {
+         return sendBeaconXHR(url, data);
+      }
+      return sendBeaconXHR(url);
+   }
+
+   // #3: fetch (ServiceWorker Scope)
+   const hasFetch = (typeof fetch !== 'undefined');
+   if (hasFetch) {
+      if (data) {
```

Comment:
**suggestion**: Do we need this condition here? It seems like `sendBeaconFetch()` can handle an undefined value for `data`, resulting in an empty body value. Any thoughts?

The same goes for [this condition](https://github.com/jahilldev/minimal-analytics/pull/45/files#diff-13e6bdc5c71c44551754344801fde739eaa74f7955cbcadd61239e44061ac056R67) and [this one](https://github.com/jahilldev/minimal-analytics/pull/45/files#diff-13e6bdc5c71c44551754344801fde739eaa74f7955cbcadd61239e44061ac056R77).

Reply:
I think I was writing to the tests, because without this the `toBeCalledWith` need to explicitly include `undefined`. From what I can tell, `jest-extended` is needed for `expect.anything()` to include `undefined` or include some other hack.

https://stackoverflow.com/questions/47027011/using-expect-anyobject-or-expect-anything-does-not-work-to-match-undefi

Reply:
Ah I see, if we know in the test case whether `data` will be undefined (or something else), couldn't we do one of the following:

```ts
// data is defined
expect(sendBeaconFetch).toBeCalledWith(expect.any(String), stubOfData);

// data is undefined
expect(sendBeaconFetch).toBeCalledWith(expect.any(String), undefined);
```

I may be missing something, but we should be able to do something similar I believe. Ideally we'd also want an explicit value for each, so rather than using `expect.any()` we pass in the stub or test value.

What do you think?